lib/charms/prometheus_k8s/v0/prometheus_scrape.py

```diff
@@ -545,84 +546,203 @@ def as_dict_with_promql_labels(self):
+        #  - suffix, from the relative path of the rule file;
+        group_name_parts = [topology.identifier, rel_path, group_name, "alerts"]
+        return "_".join(filter(None, group_name_parts))
+
+    @staticmethod
+    def _from_dir(dir_path: str, topology: JujuTopology, recursive: bool) -> List[dict]:
+        """Read all rule files in a directory.
+
+        All rules from files for the same directory are loaded into a single
+        group. The generated name of this group includes juju topology.
+        By default, only the top directory is scanned; for nested scanning, pass `recursive=True`.
+
+        Args:
+            dir_path: directory containing *.rule files (alert rules without groups).
+            topology: juju topology information.
+            recursive: flag indicating whether to scan for rule files recursively.
+
+        Returns:
+            a list of dictionaries representing prometheus alert rule groups, each dictionary
+            representing an alert group (structure determined by `yaml.safe_load`).
+        """
+        alert_groups = []  # type: List[dict]
+
+        # Gather all alerts into a list of groups
+        paths = Path(dir_path).glob("**/*.rule" if recursive else "*.rule")
+        for file_path in filter(Path.is_file, paths):
+            alert_groups_from_file = AlertRules._from_file(dir_path, file_path, topology)
+            if alert_groups_from_file:
+                logger.debug("Reading alert rule from %s", file_path)
+                alert_groups.extend(alert_groups_from_file)
 
-    paths = Path(dir_path).glob("**/*.rule" if recursive else "*.rule")  # type: ignore
-    for path in filter(Path.is_file, paths):
-        if rule := load_alert_rule_from_file(path, topology):
-            logger.debug("Reading alert rule from %s", path)
-            alerts[_group_name(path)].append(rule)
+        return alert_groups if alert_groups else []
 
-    # Gather all alerts into a list of groups since Prometheus
-    # requires alerts be part of some group
-    groups = [{"name": k, "rules": v} for k, v in alerts.items()]
-    return groups
+    def add_from_path(self, path: str, *, recursive: bool = False):
+        """Add rules from a dir path.
+
+        All rules from files are aggregated into a data structure representing a single rule file.
+        All group names are augmented with juju topology.
+
+        Args:
+            path: either a rules file or a dir of rules files.
+            recursive: whether to read files recursively or not.
+
+        Raises:
+            NotADirectoryError: if the provided path is invalid.
+
+        Returns:
+            self, so calls are chainable.
+        """
+        if Path(path).is_dir():
+            self.alert_groups.extend(AlertRules._from_dir(path, self.topology, recursive))
+        else:
+            raise NotADirectoryError("The provided path does not exist: {}".format(path))
+
+        return self  # so calls are chainable
```

Comment:
If it's possible I would use guard clauses

```python
       if not Path(path).is_dir():
            raise NotADirectoryError("The provided path does not exist: {}".format(path))
            
        self.alert_groups.extend(AlertRules._from_dir(path, self.topology, recursive))    
        return self # so calls are chainable        
```

Reply:
Agreed, but the code has changed a bit - there is an `else` now and no return.